### ISSUE TITLE
Why mmcls computes accuracy in validation step

### ISSUE THREAD

**Issue by xpngzhng** [open-mmlab/mmclassification#39]
I found some inconsistencies in mmcls, mmdet and mmseg
One major difference between mmcls and mmdet + mmseg is that mmcls requires validation dataset to collect gt_label
see https://github.com/open-mmlab/mmclassification/blob/master/configs/_base_/datasets/imagenet_bs32.py, the test pipeline
```python
test_pipeline = [
    dict(type='LoadImageFromFile'),
    dict(type='Resize', size=(256, -1)),
    dict(type='CenterCrop', crop_size=224),
    dict(type='Normalize', **img_norm_cfg),
    dict(type='ImageToTensor', keys=['img']),
    dict(type='ToTensor', keys=['gt_label']),
    dict(type='Collect', keys=['img', 'gt_label'])
]
```
and when performing validation, the forward step computes the accuracy, the dataset itself only gather the batch accuracy together to compute the final accuracy, see https://github.com/open-mmlab/mmclassification/blob/master/mmcls/datasets/base_dataset.py#L48

mmdet and mmseg compute metrics differently, in validate step, only network outputs are collected, accuracy and other metrics are computed in the standalone validation function

mmcls's test_pipeline's collecting gt_label behavior also makes the inference a little tricker than mmdet and mmseg,
I write `inference_model` in mmcls/apis/inference.py like this, notice how I modified the `test_pipeline`
```python
class LoadImage:
    """A simple pipeline to load image."""

    def __call__(self, results):
        """Call function to load images into results.

        Args:
            results (dict): A result dict contains the file name
                of the image to be read.

        Returns:
            dict: ``results`` will be returned containing loaded image.
        """

        if isinstance(results['img'], str):
            results['filename'] = results['img']
            results['ori_filename'] = results['img']
        else:
            results['filename'] = None
            results['ori_filename'] = None
        img = image.imread(results['img'])
        results['img'] = img
        results['img_shape'] = img.shape
        results['ori_shape'] = img.shape
        return results


def inference_model(model, img):
    """Inference image(s) with the classifier.

        Args:
            model (nn.Module): The loaded segmentor.
            imgs (str/ndarray or list[str/ndarray]): Either image files or loaded
                images.

        Returns:
            (list[Tensor]): The segmentation result.
        """
    cfg = model.cfg
    device = next(model.parameters()).device  # model device
    # build the data pipeline
    test_pipeline = [LoadImage()] + cfg.data.test.pipeline[1:-3] +\
        [ImageToTensor(keys=['img']), Collect(keys=['img'])]
    test_pipeline = Compose(test_pipeline)
    # prepare data
    data = dict(img=img)
    data = test_pipeline(data)
    data = collate([data], samples_per_gpu=1)
    if next(model.parameters()).is_cuda:
        # scatter to specified GPU
        data = scatter(data, [device])[0]
    else:
        if 'img_metas' in data:
            data['img_metas'] = data['img_metas'][0].data

    # forward the model
    with torch.no_grad():
        result = model(return_loss=False, **data)
    return result
```

**Comment by yl-1993** [comment]
Hi @xpngzhng , PR #16 now follows the same interface as mmdet and mmseg, which collects outputs first and then computes evaluation metric.

You're welcomed to take a review. Thanks.
